### PR TITLE
Rehearse jobs that use changed ci-operator configs

### PR DESCRIPTION
**Pull request by petr-muller**
Post-https://github.com/openshift/ci-operator-prowgen/pull/96, we are able to detect which ci-operator configs changed in a PR. To start rehearsing the jobs affected by changed ci-operator configs, we needed to implement these two remaining capabilities:

1. Walk through all presubmits and find which ones reference one of the changed configs
2. Merge the two discovered sets of rehearsal jobs without creating duplicates.

/cc @stevekuznetsov @bbguimaraes @droslean

### CODE REVIEW

File: cmd/pj-rehearse/main_test.go [hunk 1]
```diff
@@ -146,12 +147,19 @@ func getRehersalsHelper(logger *logrus.Entry, prNumber int) ([]*prowconfig.Presu
...
+		return nil, fmt.Errorf("failed to load ci-operator config: %v", err)
+	}
 
 	changedPresubmits := diffs.GetChangedPresubmits(prowConfig, prowPRConfig, logger)
 	if len(changedPresubmits) == 0 {
 		return nil, fmt.Errorf("Empty changedPresubmits was not expected")
 	}
 
+	changedCiopConfigs := diffs.GetChangedCiopConfigs(ciopMasterConfig, ciopPrConfig, logger)
+	changedPresubmits.AddAll(diffs.GetPresubmitsForCiopConfigs(prowPRConfig, changedCiopConfigs, logger))
```
Comment (stevekuznetsov) [inline cmd/pj-rehearse/main_test.go:161]: The fact that you needed to do this in the test, too, is smelly

File: pkg/config/jobs_test.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+			prowconfig.Presubmit{JobBase: prowconfig.JobBase{Name: "source-job-for-org-source-repo"}},
+		}},
+		destination: Presubmits{"org/destination-repo": {
+			prowconfig.Presubmit{JobBase: prowconfig.JobBase{Name: "destination-job-for-org-destination-repo"}}}},
+		expected: Presubmits{
+			"org/source-repo":      {prowconfig.Presubmit{JobBase: prowconfig.JobBase{Name: "source-job-for-org-source-repo"}}},
+			"org/destination-repo": {prowconfig.Presubmit{JobBase: prowconfig.JobBase{Name: "destination-job-for-org-destination-repo"}}},
+		},
+	}, {
+		description: "merge jobs with same name for a single repo, result has the one originally in destination",
```
Comment (stevekuznetsov) [inline pkg/config/jobs_test.go:60]: Since we need unique names to work this might be better as an error, but meh

File: pkg/config/release.go [hunk 3]
```diff
@@ -20,6 +20,8 @@ const (
 	CiopConfigInRepoPath = "ci-operator/config"
 	// TemplatesPath is the path of the templates from release repo
 	TemplatesPath = "ci-operator/templates"
+	// Name of the configmap that stores all ci-operator configs
+	CiOperatorConfigsCMName = "ci-operator-configs"
```
Comment (stevekuznetsov) [inline pkg/config/release.go:26]: This is not true -- we are sharded. See the `ConfigMapName()` method on the `config.Info` struct

File: pkg/rehearse/jobs.go [hunk 4]
```diff
@@ -166,7 +164,7 @@ func inlineCiOpConfig(job *prowconfig.Presubmit, targetRepo string, ciopConfigs
 
 // ConfigureRehearsalJobs filters the jobs that should be rehearsed, then return a list of them re-configured with the
 // ci-operator's configuration inlined.
-func ConfigureRehearsalJobs(toBeRehearsed map[string][]prowconfig.Presubmit, ciopConfigs config.CompoundCiopConfig, prNumber int, loggers Loggers, allowVolumes bool) []*prowconfig.Presubmit {
+func ConfigureRehearsalJobs(toBeRehearsed config.Presubmits, ciopConfigs config.CompoundCiopConfig, prNumber int, loggers Loggers, allowVolumes bool) []*prowconfig.Presubmit {
```
Comment (stevekuznetsov) [inline pkg/rehearse/jobs.go:167]: Might have been nicer to do the type changes in a separate PR

File: cmd/pj-rehearse/main_test.go (hunk 1 above)
Comment (petr-muller) [reply cmd/pj-rehearse/main_test.go:161]: True, that test basically mimics our main().

File: pkg/config/jobs_test.go (hunk 2 above)
Comment (petr-muller) [reply pkg/config/jobs_test.go:60]: It cannot be an error. In rehearse, when both ci-operator config and a presubmit that uses it change, we'll want to rehearse the same job for two reasons, so it ends up being returned by two different "what jobs do we need to rehearse" methods, and those results need to be merged.

File: pkg/config/release.go (hunk 3 above)
Comment (petr-muller) [reply pkg/config/release.go:26]: I'm aware of that, I want to tackle that tomorrow. IMHO it's worth a separate PR. I only moved this constant, not created it. Rehearse is a little broken after the sharding anyways.

File: pkg/rehearse/jobs.go (hunk 4 above)
Comment (petr-muller) [reply pkg/rehearse/jobs.go:167]: Maybe. It's basically just a rename, I did a separate commit, and it was mostly driven by the need for the `AddAll` method.

File: pkg/diffs/diffs.go [hunk 5]
```diff
@@ -141,3 +141,33 @@ func GetChangedTemplates(masterTemplates, prTemplates map[string]*templateapi.Te
...
+
+func GetPresubmitsForCiopConfigs(prowConfig *prowconfig.Config, ciopConfigs config.CompoundCiopConfig, logger *logrus.Entry) config.Presubmits {
+	ret := config.Presubmits{}
+
+	for repo, jobs := range prowConfig.JobConfig.Presubmits {
+		for _, job := range jobs {
+			if job.Agent != string(pjapi.KubernetesAgent) {
+				continue
+			}
+			for _, container := range job.Spec.Containers {
```
Comment (droslean) [inline pkg/diffs/diffs.go]: there is no need to iterate here. just use `job.Spec.Containers[0]`
There is a validation already happening in the upstream library.

File: cmd/pj-rehearse/main.go [hunk 6]
```diff
@@ -188,8 +186,10 @@ func main() {
 	}
 	loggers := rehearse.Loggers{Job: logger, Debug: debugLogger.WithField(prowgithub.PrLogField, prNumber)}
 
-	changedPresubmits := diffs.GetChangedPresubmits(masterConfig.Prow, prConfig.Prow, logger)
-	rehearsals := rehearse.ConfigureRehearsalJobs(changedPresubmits, prConfig.CiOperator, prNumber, loggers, o.allowVolumes)
+	toRehearse := diffs.GetChangedPresubmits(masterConfig.Prow, prConfig.Prow, logger)
+	toRehearse.AddAll(diffs.GetPresubmitsForCiopConfigs(prConfig.Prow, changedCiopConfigs, logger))
```
Comment (droslean) [inline cmd/pj-rehearse/main.go:203]: So, the intention here is to add all the jobs that are using the changed ci-op config?

**Review by droslean — COMMENTED** [review]
In general, this PR is ok.  But still there are some cases missing.

If there is a changed ci-op config and a changed presubmit that its using that config, there is no need to do all of this, and instead we can just rehearse the changed presubmit with the changed ci-op config inlined. I am pretty sure that in most cases this is what will happen since if someone will change his ci-op config, prowgen will generate different presubmit. 

So, the only thing left in this PR is to do this procedure only if there is no other changed presubmit that uses the changed ci-op config.

File: pkg/diffs/diffs.go (hunk 5 above)
Comment (petr-muller) [reply pkg/diffs/diffs.go]: WIll do.

File: cmd/pj-rehearse/main.go (hunk 6 above)
Comment (petr-muller) [reply cmd/pj-rehearse/main.go:203]: All the jobs that are using **some** of the changed ci-op configs, there could be more...

`AddAll` prevents job being present twice, so if a job was already in `toRehearse` because it was changed directly, it's not added again.

**Comment by petr-muller** [comment]
> If there is a changed ci-op config and a changed presubmit that its using that config, there is no need to do all of this, and instead we can just rehearse the changed presubmit with the changed ci-op config inlined. I am pretty sure that in most cases this is what will happen since if someone will change his ci-op config, prowgen will generate different presubmit.
> 
> So, the only thing left in this PR is to do this procedure only if there is no other changed presubmit that uses the changed ci-op config.

Can you elaborate on the changes that you are requesting?  We detect which jobs changed directly. Then we detect which ciop configs changed and which jobs are using them, and merge these jobs with those changed directly. The resulting set of jobs will be rehearsed. What do you think should be done differently?

**Comment by droslean** [comment]
> Can you elaborate on the changes that you are requesting? We detect which jobs changed directly. Then we detect which ciop configs changed and which jobs are using them, and merge these jobs with those changed directly. The resulting set of jobs will be rehearsed. What do you think should be done differently?

I wanted to be more selective in which jobs we should rehearse instead of running them all when we detect a changed ci-op config. I think what its implemented in this PR is a good start and we can be more selective in the future.

**Comment by petr-muller** [comment]
Addressed the actionable comments.

File: pkg/config/release.go (hunk 3 above)
Comment (petr-muller) [reply pkg/config/release.go:26]: Addressed in https://github.com/openshift/ci-operator-prowgen/pull/104

File: pkg/config/jobs_test.go (hunk 2 above)
Comment (stevekuznetsov) [reply pkg/config/jobs_test.go:60]: Ah

**Review by stevekuznetsov — APPROVED** [review]
/lgtm
